### ISSUE TITLE
Add Bytecode Outline to JDT UI

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-jdt/eclipse.jdt.ui#364]
Follow up on https://bugs.eclipse.org/bugs/show_bug.cgi?id=540436 / https://github.com/iloveeclipse/bytecodeoutline/issues/8 .

One goal would be to add "Bytecode Outline" view to JDT UI.
Stretch goal would be to replace default class file editor (that shows bytecode for classes with missing source) with one crafted by BCO.

I had no time yet to work on this, but I think that should be an easy task now.
The main issue is to move the code from https://github.com/iloveeclipse/bytecodeoutline & setup maven build in JDT UI so the bundle is picked up by platform build.

**Comment by LunarLaurus** [comment]
I believe the work done here is enough for the Maven aspect: https://github.com/LunarLaurus/eclipse.jdt.ui/tree/BytecodeOutline

My fresh Oomph install for making these changes doesn't let me launch it as an Eclipse Application (Even with a fresh clone of JDT), but I don't see why this shouldn't work.

**Comment by LunarLaurus** [comment]
I have seen that it is in the process of being approved and merged, good work.
Currently trying to get an Eclipse instance working where jdt.ui.* packages can be edited without missing dependencies so that I am ready once merge occurs. Trying to use the installer in advance mode to set up Eclipse for Eclipse developers with just the jdt.ui.* packages available.

**Comment by iloveeclipse** [comment]
See `ClassFileBytesDisassembler` for the current class file disassembler used. I think the idea of [original request](https://bugs.eclipse.org/bugs/show_bug.cgi?id=540436) was to have the output of that being replaced by properly colored / syntax highlighted bytecode (ideally with hovers, outline etc).

So I would first recommend to try out "just" making the disassembled source "nice looking". Outline & navigation are harder.

**Comment by LunarLaurus** [comment]
So I can create a new oomph intsance of for jdt packages, however while it pulls down the `org.eclipse.jdt.bcoview` repository within `git\eclipse.jdt.ui`, Eclipse itself fails to add bcoview as a project. Importing as a maven project into the jdt.ui working set works, but it still cannot find the newly added asm bundles.
![image](https://user-images.githubusercontent.com/121956871/212740660-1881fc7c-fe64-49b2-b5bf-8af21c4764d3.png)

Do I need to wait for this to deploy everywhere or? 
I will try again tomorrow, just wanted to start now 😆

**Comment by iloveeclipse** [comment]
> Do I need to wait for this to deploy everywhere or?

I assume the dependencies are **in the target platform** now (because the code builds), but not in the resulted SDK package (because the view is not referenced by any feature / product).

I assume we need same addition like in https://github.com/eclipse-platform/eclipse.platform.releng/pull/180.

I will prepare PR as soon as https://github.com/eclipse-platform/eclipse.platform.releng/pull/180 will be merged, to avoid merge conflicts.

**Comment by akurtakov** [comment]
Let me run a full I-build to be sure I haven't missed anything for the old views addition first.

**Comment by iloveeclipse** [comment]
> Let me run a full I-build to be sure I haven't missed anything for the old views addition first.

I've also noticed there is no `org.eclipse.jdt.bcoview` **feature** yet (I didn't wanted to copy old feature that has no value). 
I assume I have to create one now? I can quickly add one.

**Comment by akurtakov** [comment]
Yes, you would have to create such a feature so there is smth end user installable .

**Comment by iloveeclipse** [comment]
OK, I will add that feature then.

**Comment by LunarLaurus** [comment]
Tried to get this rolling today with yet another fresh install via Oomph, but now I get this.
Not really sure where to report this properly, but figure you do.
Sorry to be such a pain @iloveeclipse 🚧 

![image](https://user-images.githubusercontent.com/121956871/213140591-3df488d2-0b5c-4660-8e34-dc42af39c409.png)

**Comment by iloveeclipse** [comment]
Please retry once again. I've just built SDK again (https://download.eclipse.org/eclipse/downloads/drops4/I20230118-0200/), probably you've seen the version https://download.eclipse.org/eclipse/downloads/drops4/I20230117-1800/

**Comment by LunarLaurus** [comment]
Appears this still occurs even after updating Oomph. Should I just wait longer?

**Comment by iloveeclipse** [comment]
Could be again mismatch between Orbit's and maven bundle names: `org.objectweb.asm.tree.analysis` vs `org.objectweb.asm.analysis`. The later one is in Orbit, the former one is from Maven / platform.

**Comment by merks** [comment]
If the file it's complaining really does not exist on disk and the problem recurs if you try again, then probably you need to try to repair the bundle pool:

https://www.eclipse.org/forums/index.php/m/1835944/?srch=repair+bundle+pool#msg_1835944

As mentioned in that thread, sometimes virus scanners keep the file from being moved from the download location to the pool...

**Comment by LunarLaurus** [comment]
Was just in a meeting, let me poke things and then do a fresh Oomph run.
It's certainly not an AV issue though. Thank you both for being helpful!

**Comment by LunarLaurus** [comment]
I nuked my bundle pool and it redownloaded everything.
Sorry for the screenshots, but I feel it explains this issue best.

Oomph setup.
![image](https://user-images.githubusercontent.com/121956871/213196097-cbacd761-f9ab-4b1d-ad74-af0bcfb1f6db.png)
JDT UI selected for development.
![image](https://user-images.githubusercontent.com/121956871/213196138-41ad76e9-fb4b-4b14-ab32-b62ee80d5e53.png)
Oomph Okay.
![image](https://user-images.githubusercontent.com/121956871/213196445-7d98787d-d86c-4dee-b8cc-bece2497752d.png)
Eclipse Staging.
![image](https://user-images.githubusercontent.com/121956871/213196597-db5cc3ac-773b-40a9-98c1-077e8700f28d.png)
Asm Utils still missing?
![image](https://user-images.githubusercontent.com/121956871/213196654-d7aa4cfd-dd8a-403d-836d-fd8a69dcb597.png)
Version Information.
![image](https://user-images.githubusercontent.com/121956871/213196824-9df19fa0-5629-4704-a268-9c24cd540e3f.png)
ASM Utils is installed.
![image](https://user-images.githubusercontent.com/121956871/213196730-c7c77c23-ef63-4e10-a026-0d2c42b2ba28.png)
BCO Exists.
![image](https://user-images.githubusercontent.com/121956871/213196778-ce167d48-f183-46ee-8ac3-3db056aa213b.png)

**Comment by merks** [comment]
I've disabled the virus scanner from scanning my p2 folder and from scanning the folder containing all my installations.  This helps make the installations faster too!

**Comment by iloveeclipse** [comment]
> One goal would be to add "Bytecode Outline" view to JDT UI.

Done & verified with https://download.eclipse.org/eclipse/downloads/drops4/I20230118-0200

> Stretch goal would be to replace default class file editor (that shows bytecode for classes with missing source) with one crafted by BCO.

@LunarLaurus : If you still have interest to continue, please open a new ticket here with an appropriate title & description.

I think the main part is done now => closing **this** issue.

**Comment by noopur2507** [comment]
![BCO](https://user-images.githubusercontent.com/8749449/224647336-bcfb5f6f-b2d0-4139-aba0-296ecb4a8508.png)
The string "Bytecode Outline" is not translated as reported in the testing (see screenshot). @iloveeclipse, could you please check if it is externalized properly and mention the location of the externalized string here?

**Comment by merks** [comment]
Just the feature doesn't make it translatable, but that's only affects the display in p2, e.g., install/update or about:

![image](https://user-images.githubusercontent.com/208716/224648463-c7d631fb-6a93-4060-b77c-cc9a68cf3ef8.png)

Babel has not had a release in more than a year and is hoping to produce a new release for 2023-06 so until then, there is no translation available for such new strings...

**Comment by noopur2507** [comment]
Looks like the preference page string is hard-coded in /org.eclipse.jdt.bcoview/plugin.xml.

Can you please create a patch to fix it in the master and the R4_27_maintenance branches?

**Comment by iloveeclipse** [comment]
I will do, but not today. I assume this is not urgent.

**Comment by noopur2507** [comment]
You can create the patch tomorrow. It's required by IBM products team to pass 4.27 testing.

**Comment by iloveeclipse** [comment]
> You can create the patch tomorrow. 

I would try.

> It's required by IBM products team to pass 4.27 testing.

Shouldn't that be driven by IBM products team then?

Ideally it would be found during code reviews, but once code is merged - at least short after merge, not few days before release. In our company we run all the Eclipse related tests **we** have every week against latest Eclipse build, so we usually detect regressions latest in one week after merge to master.
